### PR TITLE
Upgrade to Next13

### PR DESCRIPTION
**Pull request by katiegoines**
#### Description of changes:

- upgraded Next to v13
- upgraded React to v18

### Instructions

**If this PR should not be merged upon approval for any reason, please submit as a DRAFT**

Which product(s) are affected by this PR (if applicable)?
- [ ] amplify-cli
- [ ] amplify-ui
- [ ] amplify-studio
- [ ] amplify-hosting
- [ ] amplify-libraries

Which platform(s) are affected by this PR (if applicable)?
- [ ] JS
- [ ] iOS
- [ ] Android
- [ ] Flutter
- [ ] React Native

**Please add the product(s)/platform(s) affected to the PR title**

#### Checks

- [ ] Does this PR conform to [the styleguide](https://github.com/aws-amplify/docs/blob/main/STYLEGUIDE.md)?

- [ ] Does this PR include filetypes other than markdown or images? Please add or update unit tests accordingly.

- [ ] Are any files being deleted with this PR? If so, have the needed redirects been created?

- [ ] Are all links in MDX files using the MDX link syntax rather than HTML link syntax? <br /> 
      _ref: MDX: `[link](https://link.com)` 
            HTML: `<a href="https://link.com">link</a>`_

### When this PR is ready to merge, please check the box below
- [ ] Ready to merge

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

**Review by josefaidt — DISMISSED** [review]
Hey @katiegoines this looks great! I was able to poke around the preview and confirm things are working as expected. Although I have concerns about the `domLoaded` piece it will be a larger effort to fix the hydration issues, and I do not see any negative impact with this in place compared to the current setup.

File: package.json [hunk 1]
```diff
@@ -31,17 +31,19 @@
     "copy-to-clipboard": "^3.2.1",
     "dotenv": "^16.0.0",
     "emotion": "^10.0.23",
+    "eslint-config-next": "^13.2.4",
     "extract-mdx-metadata": "^2.0.0",
     "gray-matter": "4.0.3",
     "html-entities": "^1.2.1",
     "instantsearch.js": "^4.39.1",
-    "next": "^12.1.4",
+    "module": "^1.2.5",
```
Comment (josefaidt) [inline package.json:39]: this is a node.js built-in https://nodejs.org/docs/latest-v16.x/api/module.html

```suggestion
```

File: src/components/InternalLink/index.tsx [hunk 2]
```diff
@@ -28,6 +31,7 @@ export default function InternalLink({ href, children }) {
           href = `${hrefParts[0]}/q/${filterKind}/${filterKey}#${hrefParts[1]}`;
         } else {
           href += `/q/${filterKind}/${filterKey}`;
+          useEffect(() => setUrl(href), []);
```
Comment (josefaidt) [inline src/components/InternalLink/index.tsx:34]: does this need a `useEffect` to set the href properly? Can all of this conditional logic be inside of the `useEffect`? https://legacy.reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level

parsing local storage (i.e. from `window`) should also reside inside the useEffect as `window` will not be available server-side  (during prerendering)

File: src/components/Layout/index.tsx [hunk 3]
```diff
@@ -10,7 +10,7 @@ import {
   RIGHT_NAV_LINKS,
   SOCIAL_LINKS
 } from '../../utils/globalnav';
-import React from 'react';
+import React, { useEffect, useState } from 'react';
```
Comment (josefaidt) [inline src/components/Layout/index.tsx:13]: this isn't entirely necessary to change but the `React` import is no longer required for React 17+ https://legacy.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html

File: src/components/SearchBar/Autocomplete.tsx [hunk 4]
```diff
@@ -98,8 +100,14 @@ export function Autocomplete(props) {
...
-        render(children, root);
-      },
+        if (!panelRootRef.current || rootRef.current !== root) {
+          rootRef.current = root;
+
+          panelRootRef.current?.unmount();
+          panelRootRef.current = createRoot(root);
+        }
+
+        panelRootRef.current.render(children);      },
```
Comment (josefaidt) [inline src/components/SearchBar/Autocomplete.tsx:110]: what is this doing? Is it possible to use `React.lazy` with Suspense here? https://legacy.reactjs.org/docs/code-splitting.html#reactlazy

or maybe `next/dynamic`? https://nextjs.org/docs/advanced-features/dynamic-import

Comment (timngyn) [reply src/components/SearchBar/Autocomplete.tsx:110]: I'm also a little confused by this as well. Should we add some comments for why we're doing this?

**Review by timngyn — COMMENTED** [review]
It looks good so far! Thank you for doing this @katiegoines

I have similar questions to what @josefaidt asked as well with regards to the `useEffect` for calling `setUrl`

File: src/components/Menu/FilterSelect/index.tsx [hunk 5]
```diff
@@ -31,30 +31,19 @@ const convertToRouteHerf = (
...
-  return {
-    pathname: path,
-    query: {
-      [filter.filterKind]: targetFilterKey,
-      firstPathSegment: multiLibVersionPlatforms.includes(targetFilterKey)
-        ? firstPathSegment
-        : firstPathSegment?.split('-')[0]
-    }
-  };
+  return path;
```
Comment (jacoblogan) [inline src/components/Menu/FilterSelect/index.tsx:46]: It looks like here we are changing the return type of this function but not how we use it.  Is the expected output for this ever expected to be the object that was previously being returned?

File: src/components/Callout/styles.tsx [hunk 6]
```diff
@@ -18,7 +18,7 @@ export const CalloutOuter = styled.div<CalloutProps>((props) => {
     margin-bottom: 0;
     margin-top: 1rem;
 
-    &:first-child {
+    &:first-of-type {
```
Comment (jacoblogan) [inline src/components/Callout/styles.tsx:21]: These two selectors will often have the same effect but not always.  Just for my own understanding, why the switch to first-of-type?

File: src/components/SearchBar/Autocomplete.tsx (hunk 4 above)
Comment (katiegoines) [reply src/components/SearchBar/Autocomplete.tsx:110]: This is following the Algolia recommendation for upgrading to React18. Looking into the alternatives suggested.

File: src/components/InternalLink/index.tsx (hunk 2 above)
Comment (katiegoines) [reply src/components/InternalLink/index.tsx]: This is hydration related. I missed removing it when separating out the hydration efforts. Will address this on that branch. Thanks

File: src/components/Menu/FilterSelect/index.tsx [hunk 7]
```diff
@@ -106,7 +101,7 @@ export default class FilterSelect extends React.Component<
     }
 
     return (
-      <Link href={href} key={name}>
+      <Link href={href} key={name} legacyBehavior>
```
[the remark this answers is not in the document]
Comment (katiegoines) [reply src/components/Menu/FilterSelect/index.tsx:98]: Yes, in some places it did not make sense to remove out the <a> when there is the workaround.

File: src/components/Callout/styles.tsx (hunk 6 above)
Comment (katiegoines) [reply src/components/Callout/styles.tsx]: We were getting a warning, but it's not related to the upgrade, so I've removed these and they can be addressed at another time.

**Review by josefaidt — APPROVED** [review]
LGTM 🚀 Great work on this, @katiegoines!

**Review by jacoblogan — APPROVED** [review]

**Review by jacoblogan — APPROVED** [review]

**Review by josefaidt — APPROVED** [review]

**Pull request merged**